### PR TITLE
fix: restore space widths on partial cache misses

### PR DESCRIPTION
**Pull request by serialx**
## Summary

SD font spaces can collapse to zero width when the advance cache is full or only partially populated across styles. Words then run together and line breaks move, even though the font has a valid space glyph.

`hasAdvanceTable()` only means that some style has a table, while `getAdvance()` returns zero for both a missing entry and a genuine zero advance. `getSpaceWidth()` and `getSpaceAdvance()` treated that zero as a cache hit.

Treat a zero advance for the space codepoint as a miss and fall back to the resolved style's space glyph, the same rule the per-codepoint slow path in `getTextAdvanceX()` already applies. Both methods use the same 12.4 fixed-point rounding. The change is one helper in `GfxRenderer.cpp`: 14 added lines, four removed.

Related to #2002 and #2157. @uxjulia previously proposed space fallbacks in #2634, ported from CrossInk. That PR closed unmerged after the original overlapping-text report could no longer be reproduced; the discussion identified the existing `getTextAdvanceX()` fallback. This PR addresses the remaining gap in the two space methods with a focused reproduction. It leaves cache allocation strategy, formats, and existing SD layout kerning behavior unchanged.

No host tests are added; the fix touches only the renderer, and verification is the device reproduction below. The exhaustive audit harness is kept outside this PR.

## Scope Check

- [x] I have read SCOPE.md and ROADMAP.md.
- [x] This PR is not a new built-in theme.
- [x] This PR is not a new external network connector.
- [x] This PR is not an interactive app, writing tool, RSS/news/browser, media playback, or PDF feature.
- [x] This fixes an existing CrossPoint regression in core text layout; stock/fork feature coverage does not remove the need for the fix.
- [x] No SDK, HAL, bootloader, OTA, or recovery code changes.

### Device reproduction

Tested on an Xteink X3 with the same #3510 serial-control patch and diagnostic code on both builds. A controlled reproduction fills the advance table without space, then lays out the same 36 words through the production renderer using Pretendard 16 pt and the same available width:

| | Baseline | This change |
|---|---:|---:|
| Space width and advance | 0 px | 8 px |
| Rendered lines | 6 | 7 |

#### Before / after screenshots

Device framebuffer captures of the controlled full-cache reproduction described above.

| Before — baseline, 0 px spaces / 6 lines | After — this fix, 8 px spaces / 7 lines |
|---|---|
| before | after |

The device probe also verifies missing-style coverage, style fallback, and an injected metrics-preparation failure. The failure is deliberately injected; natural device OOM behavior has not been established by this test. These are controlled edge cases; the normal-book runs did not reproduce missing spaces.

### Measurements

Baseline `b0b844fc` versus that baseline plus this change, both with identical test instrumentation. Following the protocol in #3569: 데미안, Pretendard 16 pt, portrait, justified, Extra Wide, anti-aliased text, default spacing. Three alternating boots per build; each deleted the book cache, reopened at spine 3/page 1, and made six warmup and six measured turns over the same pages: **18 measured turns per build**. The initial setup pair is excluded because its startup history differed.

| Measurement | Baseline | This change | Difference |
|---|---:|---:|---:|
| Median page drawing (`bw_render`) | 75.0 ms | 76.5 ms | +1.5 ms |
| Median total rendering, including refresh | 1,417 ms | 1,425 ms | +8 ms |
| Median cold layout miss to first completed page | 4,472 ms | 4,489 ms | +17 ms |
| Median free heap after measured pages | 49,372 B | 49,408 B | +36 B |
| Median largest free block after measured pages | 22,516 B | 24,052 B | +1,536 B |
| Lowest heap watermark through measured pages | 23,076 B | 23,076 B | 0 B |
| Static RAM | 56,688 B | 56,688 B | 0 B |
| Linked flash | 5,565,689 B | 5,565,759 B | +70 B |

These are observations from one device/workload; the small differences do not establish a general speed or heap improvement. The fix adds no allocation sites, persistent buffers, or object fields. A miss uses the existing glyph cache, including its existing checked bitmap allocation if the space glyph has bitmap data. The full-cache probe measured about 8.3 ms for the first corrected width call, which now performs the necessary SD lookup.

Normal-book reading-area screenshots matched across builds. Existing rejected 10/12 pt UI font files affected both builds; the tested 16 pt reader font loaded successfully.

### Validation

- The existing host test suite passes unchanged.
- Extended local validation, archived outside this PR, also covered a full cache missing space, four allocation-failure points, style resolution, rounding boundaries, and representative line breaks. Five of those seven cases failed before the fix; all seven passed afterward, including under AddressSanitizer and UndefinedBehaviorSanitizer. LeakSanitizer was unavailable on the macOS host.
- The default firmware build, repository formatting, and diff checks pass.

Allocation interception, the full-cache generator, and the wrapping simulation are excluded from the PR, along with test-only serial controls and device probes. Broader font/book coverage and longer sessions under natural memory pressure remain to be tested.

### AI Usage

Did you use AI tools to help write this code? **YES**

### CODE REVIEW

File: test/sd_card_font/SdCardFontTest.cpp [hunk 1]
```diff
@@ -214,3 +214,51 @@ TEST(SdCardFontTest, BitmapAllocationRetriesAfterEvictingRebuildableAdvances) {
...
+
+TEST(SdCardFontTest, TryGetAdvanceReportsCachedZeroAsHit) {
+  makeSpaceOnlyFont();
+  SdCardFont font;
+  ASSERT_TRUE(font.load("fixture"));
+  ASSERT_EQ(0, font.buildAdvanceTable(" ", 1 << 1));
+  uint16_t advance = 0xFFFF;
+  EXPECT_TRUE(font.tryGetAdvance(' ', 1, advance));
+  EXPECT_EQ(0, advance);
+}
```
[the remark this answers is not in the document]
Comment (serialx) [reply test/sd_card_font/SdCardFontTest.cpp]: Deliberately omitted. A renderer-level test means compiling all of `GfxRenderer` on the host against a hand-maintained `HalDisplay` stub plus linker dead-strip setup, which is what the previous revision did and what this commit removed to keep the test footprint proportional to a 25-line fix. The miss/zero conflation lives in `SdCardFont`, and that is what the two host tests cover. The renderer fallback is four lines of null-checked glyph lookup and is verified by the device reproduction in the PR description.

**Comment by Uri-Tauber** [comment]
I dont understand what it came to solve.
How many glyphs have zero width? Sounds like malformed font file.

EDIT: nmv. I get it now. In a rare case when the space was absent from that style’s cache, the renderer interpreted a cache miss as a zero-pixel space, so words touch one another.

**Comment by Uri-Tauber** [comment]
Can we simply treat a zero result for `U+0020` as a miss and load the glyph?

**Comment by serialx** [comment]
Yes, and that is the better shape. Pushed; the branch now differs from develop in `GfxRenderer.cpp` only (b4af4577).

The per-glyph slow path in `getTextAdvanceX()` already treats a zero advance as "not cached, read the glyph". The two space methods were the only callers that did not. The fix is now a single helper in `GfxRenderer.cpp` that does the same for U+0020; the `SdCardFont` API addition and its tests are gone.

**Review by Uri-Tauber — APPROVED** [review]
Awesome
